### PR TITLE
Add buildtest commands option

### LINKED ISSUE #1509: [FEATURE]: buildtest commands
### Please describe your feature

Let's add an option called `buildtest commands` which will show an output of  buildtest commands as a raw output. So in the `--help` we see these commands

```
COMMANDS:
  
    build (bd)          Build and Run test
    buildspec (bc)      Buildspec Interface
    config (cg)         Query buildtest configuration
    report (rt)         Query test report
    inspect (it)        Inspect a test based on NAME or ID
    path                Show path attributes for a given test
    history (hy)        Query build history
    schema              List schema contents and examples
    cdash               Upload test to CDASH server
    unittests (test)    Run buildtest unit tests
    stylecheck (style)  Run buildtest style checks
    cd                  change directory to root of test given a test name
    clean               Remove all generate files from buildtest including
                        test directory, log files, report file, buildspec
                        cache, history files.
    docs                Open buildtest docs in browser
    schemadocs          Open buildtest schema docs in browser
    debugreport (debug)
                        Display system information and additional information
                        for debugging purposes.
    stats               Show test statistics for given test
    info                Show details regarding current buildtest setup
    help (h)            buildtest command guide
    tutorial-examples   Generate documentation examples for Buildtest Tutorial
```

The output should be something like

```
$ buildtest commands
build
buildspec
config
report
inspect
path
history
schema
cdash
unittests
stylecheck
cd
clean
docs
schemadocs
debugreport
stats
info
help
tutorial-examples
```

# TODO

- [x] Add command option in https://github.com/buildtesters/buildtest/blob/devel/buildtest/cli/__init__.py
- [x] Add to [bash completion](https://github.com/buildtesters/buildtest/blob/devel/bash_completion.sh) script
- [x] Add implementation perhaps create a module `buildtest/cli/commands.py`
- [ ] Add regression test perhaps create a separate file in `tests/cli/test_commands.py`
- [ ] Add documentation let's put this in https://buildtest.readthedocs.io/en/devel/features.html

### Suggest potential solution

_No response_

### Additional Information

_No response_

### Post question in Slack

- [X] I agree that I posted my question in [slack](https://hpcbuildtest.slack.com/) before creating this issue

### Is there an existing issue

- [X] I confirm there is no existing [issue](https://github.com/buildtesters/buildtest/issues) for this issue

### PR DESCRIPTION
**Pull request by braeeast2001**
Create a feature to show an output of Buildtest commands, by doing the following:

> - [x] 1. Adding a command option in [buildtest/cli/__init__.py](https://github.com/buildtesters/buildtest/blob/devel/buildtest/cli/__init__.py) 
> - [x]  2. Adding to [bash completion](https://github.com/buildtesters/buildtest/blob/devel/bash_completion.sh) script -
> - [x] 3. Adding implementation; creating a module buildtest/cli/commands.py 
> - [x] 4. Adding a regression test; creating a separate file in tests/cli/test_commands.py

### CODE REVIEW

**Comment by braeeast2001** [comment]
@shahzebsiddiqui When I run the command "buildtest help commands" the output I sent to you in Slack appears. Is this what I am supposed to be fixing?

**Comment by shahzebsiddiqui** [comment]
> @shahzebsiddiqui When I run the command "buildtest help commands" the output I sent to you in Slack appears. Is this what I am supposed to be fixing?

No that not what i asked for i want you to add a command `buildtest commands` that will display all the subcommands. See https://github.com/buildtesters/buildtest/issues/1509#issue-1782995433

File: buildtest/cli/commands.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    table.add_row("clean")
+    table.add_row("docs")
+    table.add_row("schemadocs")
+    table.add_row("debugreport")
+    table.add_row("stats")
+    table.add_row("info")
+    table.add_row("help")
+    table.add_row("tutorial-examples")
+
+    console.print(table)
```
Comment (shahzebsiddiqui) [inline buildtest/cli/commands.py]: @braeeast2001 can you please print this as raw text and not as table. Just a for loop and use `print`. Each item should be in new line so output should look like 

```
$ buildtest commands
build
buildspec
config 
...
```

Comment (braeeast2001) [reply buildtest/cli/commands.py]: Done!

**Comment by shahzebsiddiqui** [comment]
@braeeast2001 i fixed your PR and rewrote this to be easier to understand. Please add the regression test and we can merge this in.

Lets not document this feature for now its not necessary.

File: tests/cli/test_commands.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import pytest
+from buildtest.cli.commands import commands_cmd
+
+
+def test_commands_cmd():
+    with pytest.raises(SystemExit):
```
Comment (shahzebsiddiqui) [inline tests/cli/test_commands.py]: @braeeast2001 why are you catching exception this should run without exception

File: tests/cli/test_commands.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from buildtest.cli.commands import commands_cmd
+
+
+def test_commands_cmd():
```
Comment (shahzebsiddiqui) [inline tests/cli/test_commands.py]: can you name this method `test_buildtest_commands`

File: buildtest/cli/commands.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+def commands_cmd():
```
Comment (shahzebsiddiqui) [inline buildtest/cli/commands.py]: can you rename this method and all instances where this is called to `list_buildtest_commands`

**Comment by shahzebsiddiqui** [comment]
@braeeast2001 can please look at the CI checks and fix them

**Comment by shahzebsiddiqui** [comment]
@braeeast2001 can you please black your code. Look at the style check it reports 3files need formatting.

Just run `buildtest style -a`

File: bash_completion.sh [hunk 5]
```diff
@@ -456,8 +456,8 @@ _buildtest ()
       COMPREPLY=( $( compgen -W "${cmds}" -- $cur ) )
       ;;
     *)
-      local cmds="build buildspec cd cdash clean config debugreport docs help info inspect history path report schema schemadocs stats stylecheck tutorial-examples unittests"
-      local alias_cmds="bd bc cg debug it h hy rt style test"
+      local cmds="build buildspec cd cdash clean config debugreport docs help info inspect history path report schema schemadocs stats stylecheck tutorial-examples unittests commands"
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: This list needs to be in alphabetical order so put `commands` in its right place

File: bash_completion.sh [hunk 6]
```diff
@@ -456,8 +456,8 @@ _buildtest ()
...
+      local alias_cmds="bd bc cg debug it h hy rt style test cmd"
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: Same here please alphabetized this list.

File: buildtest/cli/commands.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+def list_buildtest_commands():
+    """This method implements command ``buildtest commands`` which shows a list
```
Comment (shahzebsiddiqui) [inline buildtest/cli/commands.py]: Let's just make this one line. I see this statement broken up in 3 lines. The sentence is not that long for it to require new line.

**Comment by shahzebsiddiqui** [comment]
@braeeast2001 can you please recreate your python environment and reinstall buildtest it looks like your black and isort are not using the configuration specified from pyproject.toml

This might address some formatting issues you are facing.

Please delete your existing python virtual environment.

You should consider using pipenv it will simplify managing virtual environment

File: buildtest/cli/commands.py (hunk 7 above)
Comment (braeeast2001) [reply buildtest/cli/commands.py]: I got a "weak warning" about this previously, I am correcting this now.

File: bash_completion.sh [hunk 8]
```diff
@@ -456,9 +456,9 @@ _buildtest ()
       COMPREPLY=( $( compgen -W "${cmds}" -- $cur ) )
       ;;
     *)
-      local cmds="build buildspec cd cdash clean config debugreport docs help info inspect history path report schema schemadocs stats stylecheck tutorial-examples unittests commands"
-      local alias_cmds="bd bc cg debug it h hy rt style test cmd"
-      local opts="--color --config --debug --editor --help --helpcolor --logpath --loglevel --print-log --no-color --report --version --view-log -c -d -h -l -p -r -V"
+      local cmds="build buildspec cd cdash clean commands config debugreport docs help history info inspect path report schema schemadocs stats stylecheck tutorial-examples unittests"
+      local alias_cmds="bd bc cg cmd debug h hy it rt style test"
+      local opts="--color --config --debug --editor --help --helpcolor --loglevel --logpath --no-color --print-log --report --version --view-log -c -d -h -l -p -r -v"
```
Comment (shahzebsiddiqui) [inline bash_completion.sh]: its a `-V` not `-v` in the help output, can you please fix this. Thanks for sorting these

```
  -V, --version         show program's version number and exit

```

File: buildtest/cli/commands.py [hunk 9]
```diff
@@ -1,8 +1,7 @@
-def list_buildtest_commands():
-    """This method implements command ``buildtest commands`` which shows a list
-    of buildtest commands"""
+def list_buildtest_commands(): """This method implements command ``buildtest commands`` which shows a list of buildtest commands"""
```
Comment (shahzebsiddiqui) [inline buildtest/cli/commands.py]: put a new line after the method it looks like docstring is on same line as the method declaration. Something like this. Make sure its indented properly

```python
def list_buildtest_commands(): 
    """This method implements command ``buildtest commands`` which shows a list of buildtest commands"""
```

File: buildtest/cli/commands.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    "debugreport",
+    "docs",
+    "history",
+    "info",
+    "inspect",
+    "path",
+    "report",
+    "schema",
+    "schemadocs",
+    "show" "stats",
```
Comment (shahzebsiddiqui) [inline buildtest/cli/commands.py]: there should be a newline and a comma. Please fix this @braeeast2001

**Comment by braeeast2001** [comment]
Change completed, ready to merge.

**Comment by shahzebsiddiqui** [comment]
@braeeast2001 can you please resolve the conflict

**Review by shahzebsiddiqui — APPROVED** [review]
thanks @braeeast2001 this looks good.

**Pull request merged**